javascript/ql/src/semmle/javascript/NullSensitiveContext.qll

```diff
@@ -0,0 +1,488 @@
+/**
+ * Provides classes for working with contexts sensitive to containing null
+ * and undefined.
+ */
+
+import javascript
+
+/**
+ * A `NullSensitiveContext` is a syntactic context in which neither null nor
+ * undefined should occur, because the context is guaranteed to inspect its
+ * content and thereby throw an error or produce other undesirable behavior.
+ *
+ * The expression found in that context is called the principal argument of
+ * the context.
+ *
+ * In examples, we denote the position of the principal argument by use of
+ * an underscore. For example, in `_ + 0`, the context's principle argument
+ * is the left operand of the addition.
+ */
+abstract class NullSensitiveContext extends Expr {
+  abstract Expr getPrincipalArgument();
+}
+
+/*
+ * _[p]
+ */
+
+class IndexExprBase extends NullSensitiveContext, IndexExpr {
+  override Expr getPrincipalArgument() { result = this.getBase() }
+}
+
+/*
+ * o[_]
+ */
+
+class IndexExprIndex extends NullSensitiveContext, IndexExpr {
+  override Expr getPrincipalArgument() { result = this.getPropertyNameExpr() }
+}
+
+/*
+ * _.p
+ */
+
+class DotExprBase extends NullSensitiveContext, DotExpr {
+  override Expr getPrincipalArgument() { result = this.getBase() }
+}
+
+/*
+ * new _
+ */
+
+class NewExprCallee extends NullSensitiveContext, NewExpr {
+  override Expr getPrincipalArgument() { result = this.getCallee() }
+}
+
+/*
+ * _(args)
+ */
+
+class CallExprCallee extends NullSensitiveContext, CallExpr {
+  override Expr getPrincipalArgument() { result = this.getCallee() }
+}
+
+/*
+ * _ + y
+ */
+
+class AddExprLeftOperand extends NullSensitiveContext, AddExpr {
```

Comment:
You can simplify the huge list of subclasses below by generalising to `BinaryExpr` and excluding `LogicalBinaryExpr` and `EqualityTest` in the characteristic predicate.

Reply:
And you also don't need separate subclasses for the left and right argument if you generalise `getPrincipalArgument` to `getASensitiveArgument` and have it return both arguments.

Reply:
I suppose for the purposes I wrote it, I can collapse things a bunch and folx can then just use additional casting or instance predicates to do more detailed case analysis.